src/Build.UnitTests/Graph/ParallelWorkSet_Tests.cs

```diff
@@ -0,0 +1,253 @@
+                new Dictionary<string, string>(StringComparer.Ordinal);
+            internal bool ShouldExpectException { get; set; }
+        }
+
+        private struct WorkItem
+        {
+            internal string Key { get; set; }
+
+            internal Func<string> WorkFunc { get; set; }
+        }
+
+        private ParallelWorkSet<string, string> _workSet;
+
+        [Fact]
+        public void GivenExceptionsOutsideWorker_CompletesAndThrowsException()
+        {
+            TestParallelWorkSet(new ParallelWorkSetTestCase
+            {
+                DegreeOfParallelism = 0,
+                WorkItemsToAdd = new List<WorkItem>
+                {
+                    new WorkItem
+                    {
+                        Key = "fooKey",
+                        WorkFunc = () => throw new Exception()
+                    },
+                    new WorkItem
+                    {
+                        Key = "barKey",
+                        WorkFunc = () => throw new Exception()
+                    },
+                    new WorkItem
+                    {
+                        Key = "bazKey",
+                        WorkFunc = () => throw new Exception()
+                    }
+                },
+                ShouldExpectException = true
+            });
+        }
+
+        [Fact]
+        public void GivenExceptionsWithinWorker_CompletesAndThrowsExceptions()
+        {
+            TestParallelWorkSet(new ParallelWorkSetTestCase
+            {
+                DegreeOfParallelism = Environment.ProcessorCount,
+                WorkItemsToAdd = new List<WorkItem>
+                {
+                    new WorkItem
+                    {
+                        Key = "fooKey",
+                        WorkFunc = () => throw new Exception()
+                    },
+                    new WorkItem
+                    {
+                        Key = "barKey",
+                        WorkFunc = () => throw new Exception()
+                    },
+                    new WorkItem
+                    {
+                        Key = "bazKey",
+                        WorkFunc = () => throw new Exception()
+                    }
+                },
+                ShouldExpectException = true
+            });
+        }
+
+        [Fact]
+        public void GivenNoWorkItemAndMultipleWorkers_Completes()
+        {
+            TestParallelWorkSet(new ParallelWorkSetTestCase
+            {
+                DegreeOfParallelism = Environment.ProcessorCount
+            });
+        }
+
+        [Fact]
+        public void GivenNoWorkItemAndNoWorkers_Completes()
+        {
+            TestParallelWorkSet(new ParallelWorkSetTestCase());
+        }
+
+        [Fact]
+        public void GivenRecursiveWorkItemsAndMultipleWorkers_Completes()
+        {
+            TestParallelWorkSet(new ParallelWorkSetTestCase
+            {
+                DegreeOfParallelism = Environment.ProcessorCount,
```

Comment:
Should these be `Environment.ProcessorCount`, or hard-coded to some number to catch the scenario where the tests are run on a single-proc VM?

Reply:
Maybe a hard-coded number if on a single-proc or below some threshold, else ```Environment.ProcessorCount```?

Reply:
Sounds like it's time for a `[Theory]` :)